### PR TITLE
Update no_trainer scripts with new Accelerate functionalities

### PR DESCRIPTION
**Pull request by muellerzr**
# Update the `no_trainer` scripts to keep aligned with Accelerate capabilities

## What does this add?

Updates all `no_trainer` scripts to use the latest capabilities.

## Why is it needed?

Accelerate had a number of new capabilities added, including better saving/loading, experiment tracking, and support for LR Schedulers. As a result, much of the current scripts can either be simplified from their hard-coded behaviors, or have these features added

Modified scripts with potential major changes:
- `language-modeling`
- `multiple-choice`
- `question-answering`
- `summarization`
- `text-classification`
- `token-classification`
- `translation`

The speech fine-tuning will be updated in a later PR

## Basic usage examples:

* Saving checkpoints each epoch or number of steps:
```bash
accelerate launch language-modeling/run_clm_no_trainer --checkpointing_steps "epoch"
```
```bash
accelerate launch language-modeling/run_clm_no_trainer --checkpointing_steps 100
```
* Resuming training from a saved checkpoint: 
```bash
accelerate launch language-modeling/run_clm_no_trainer --resume_from_checkpoint "epoch_1"
```
* Use any available trackers that Accelerate can automatically pick up including Weights and Biases, TensorBoard, and CometML
```bash
accelerate launch language_modeling/run_clm_no_trainer --with_tracking
```
## Anticipated maintence burden? (What will happen in say, 3 months if something changes)

As it gets more widly used, these scripts might need small updates if we find the end-users prefer a different experience when it comes to logging, or other small bugfixes we find as time goes on.

### CODE REVIEW

File: examples/pytorch/language-modeling/run_clm_no_trainer.py [hunk 1]
```diff
@@ -208,7 +225,11 @@ def main():
     args = parse_args()
 
     # Initialize the accelerator. We will let the accelerator handle device placement for us in this example.
-    accelerator = Accelerator()
+    # If we're using tracking, we also need to initialize it here and it will pick up all supported trackers in the environment
+    if args.with_tracking:
+        accelerator = Accelerator(log_with="all")
+    else:
+        accelerator = Accelerator()
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm_no_trainer.py:232]:
```suggestion
    accelerator = Accelerator(log_with="all") if args.with_tracking else Accelerator()
```
Short enough to be in one line I think.

File: examples/pytorch/language-modeling/run_clm_no_trainer.py [hunk 2]
```diff
@@ -427,16 +448,11 @@ def group_texts(examples):
...
-        model, optimizer, train_dataloader, eval_dataloader
-    )
-
     # On TPU, the tie weights in our model have been disconnected, so we need to restore the ties.
     if accelerator.distributed_type == DistributedType.TPU:
         model.tie_weights()
 
-    # Note -> the training dataloader needs to be prepared before we grab his length below (cause its length will be
+    # Note -> the training dataloader needs to be prepared before we grab its length below (cause its length will be
     # shorter in multiprocess)
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm_no_trainer.py:456]: That comment should now be removed since we pass the scheduler to `prepare`.

File: examples/pytorch/language-modeling/run_clm_no_trainer.py [hunk 3]
```diff
@@ -467,11 +500,39 @@ def group_texts(examples):
...
+            dirs = [f.name for f in os.scandir(os.getcwd()) if f.is_dir()]
+            dirs.sort(key=os.path.getctime)
+            path = dirs[-1]  # Sorts folders by date modified, most recent checkpoint is the last
+            if "epoch" in path.name:
+                num_epochs -= int(path.name.replace("epoch_", ""))
+            else:
+                resume_step = int(path.name.replace("step_", ""))
+                num_epochs -= resume_step // len(train_dataloader)
+                resume_step = (num_epochs * len(train_dataloader)) - resume_step
+                state_restored = False
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm_no_trainer.py:521]: This all needs to be done in the first branch as well!

**Review by sgugger — APPROVED** [review]
Thanks for working on those examples! Left a couple of comments, but looks pretty good already!

**Review by sgugger — APPROVED** [review]
Thanks for duplicating the effort across all the examples!

**Comment by Ericmututu** [comment]
Hi,
When I enable `--with_tracking` within `run_glue_no_trainer.py`, nothing seems to happen. After the training is over, I don't find any log files in `output_dir`. How do I save these training results (e.g., accuracy, training_loss, ...) as shown in the following figure?
image

Thanks in advance!
